src/OpenTelemetry/Metrics/ThreadStaticStorage.cs

```diff
@@ -49,30 +49,44 @@ internal static ThreadStaticStorage GetStorage()
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> tags, out string[] tagKeys, out object[] tagValues)
+        internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> tags, int maxLength, HashSet<string> tagKeysInteresting, out string[] tagKeys, out object[] tagValues, out int actualLength)
         {
-            var len = tags.Length;
-
-            if (len == 0)
+            if (maxLength <= MaxTagCacheSize)
+            {
+                tagKeys = this.tagStorage[maxLength - 1].TagKey;
+                tagValues = this.tagStorage[maxLength - 1].TagValue;
+            }
+            else
             {
-                throw new InvalidOperationException("There must be atleast one tag to use ThreadStaticStorage.");
+                tagKeys = new string[maxLength];
+                tagValues = new object[maxLength];
             }
 
-            if (len <= MaxTagCacheSize)
+            if (tagKeysInteresting != null)
```

Comment:
Maybe we can step back and put some design principle like this - if the user does not specify a custom list of tags, we should spend absolute zero CPU cycle related to the `tagKeysInteresting` checks.

Reply:
Pushed some changes.. close to it, but still not absolute zero. We do one if() check.